Python/rerf/urerf.py

```diff
@@ -0,0 +1,252 @@
+import multiprocessing
+
+import numpy as np
+from scipy.sparse import csr_matrix
+from sklearn.base import BaseEstimator
+from sklearn.utils import check_array
+from sklearn.utils.validation import check_is_fitted
+
+import pyfp
+
+
+class UnsupervisedRandomForest(BaseEstimator):
+    """Unsupervised random(er) forest
+
+    Supports both Random Forest, developed by Breiman (2001) [#Breiman]_, as well as
+    Randomer Forest or Random Projection Forests (RerF) developed by
+    Tomita et al. (2016) [#Tomita]_.
+
+    The difference between the two algorithms is where the random linear
+    combinations occur: Random Forest combines features at the tree level
+    whereas RerF combines features at the node level.
+
+    In addition to the normal RandomForestClassifier parameters, there are
+    two parameters to be aware of:
+
+        - ``projection_matrix``
+        - ``feature_combinations``
+
+    Parameters
+    ----------
+    projection_matrix : str, optional (default: "RerF")
+        The random combination of features to use: either "RerF", "Base".
+        "RerF" randomly combines features for each `mtry`. Base is our
+        implementation of Random Forest. "S-RerF" is structured RerF,
+        combining multiple features together in random patches.
+        See Tomita et al. (2016) [#Tomita]_ for further details.
+    n_estimators : int, optional (default: 100)
+        Number of trees in forest.
+    max_depth : int or None, optional (default=None)
+        The maximum depth of the tree. If None, then nodes are expanded
+        until all leaves are pure or until all leaves contain less than
+        min_samples_split samples.
+    min_samples_split : int, optional (default: 1)
```

Comment:
I thought we had determined default of 1 resulted in poor performance.
However, its current default is "auto"